core_controller/src/control_loop.rs

```diff
@@ -0,0 +1,262 @@
+use crate::config;
+use crate::drivers::{
+    core_driver::{CoreDriver, CoreMetadata},
+    docker::DockerDriver,
+    kubernetes::KubernetesDriver,
+    mq::{Queue, RabbitMQDriver},
+};
+use std::fmt::Debug;
+use std::{
+    fmt::{self, Formatter},
+    process,
+    sync::{
+        atomic::{AtomicBool, Ordering},
+        Arc,
+    },
+    time::Duration,
+};
+use tokio::sync::Mutex;
+use tokio::time::sleep;
+use tracing::{error, info, warn};
+
+pub struct ControlLoop {
+    pub config: config::Config,
+    pub should_continue: Arc<AtomicBool>,
+    pub known_cores: Arc<Mutex<Vec<CoreMetadata>>>,
+
+    core_driver: Box<dyn CoreDriver>,
+    mq_driver: RabbitMQDriver,
+    redis_client: redis::Client,
+}
+
+impl Debug for ControlLoop {
+    fn fmt(&self, f: &mut Formatter<'_>) -> fmt::Result {
+        write!(
+            f,
+            "ControlLoop (exiting: {})",
+            !self.should_continue.load(Ordering::Relaxed)
+        )
+    }
+}
+
+impl ControlLoop {
+    /// Create a new control loop
+    pub async fn new(
+        config: config::Config,
+        known_cores: Arc<Mutex<Vec<CoreMetadata>>>,
+    ) -> ControlLoop {
+        ControlLoop {
+            config: config.clone(),
+            core_driver: match config.provider {
+                config::Provider::Docker(options) => {
+                    info!("Using Docker driver");
+                    Box::new(DockerDriver::new(options))
+                }
+                config::Provider::Kubernetes(options) => {
+                    info!("Using Kubernetes driver");
+                    Box::new(KubernetesDriver::new(options).await)
+                }
+            },
+            mq_driver: RabbitMQDriver::new(config.rabbitmq),
+            redis_client: redis::Client::open(config.redis.url)
+                .expect("Failed to open redis connection"),
+            should_continue: Arc::new(AtomicBool::new(true)),
+            known_cores: known_cores.clone(),
+        }
+    }
+
+    /// Remove stale cores
+    #[tracing::instrument(skip_all)]
+    async fn remove_stale_cores(
+        &self,
+        cores: Vec<CoreMetadata>,
+    ) -> Result<Vec<CoreMetadata>, Box<dyn std::error::Error>> {
+        // Connect to redis
+        let mut redis_conn = match self.redis_client.get_multiplexed_tokio_connection().await {
+            Ok(conn) => conn,
+            Err(e) => {
+                warn!("Failed to connect to redis, {}", e);
+                return Err("error connecting to redis".into());
+            }
+        };
+
+        let mut remaining_cores: Vec<CoreMetadata> = vec![];
+
+        // For each core, check when the last message was seen on redis and if it is older than the timeout, delete the core and its queue
+        for core in cores {
+            let last_seen: Option<i64> = redis::cmd("GET")
+                .arg(format!(
+                    "{}/{}",
+                    &self.config.redis.core_last_seen_prefix, core.infra_id
```

Comment:
I'm suggesting to use `.` for consistency with editoast redis keys.
```suggestion
                    "{}.{}",
                    &self.config.redis.core_last_seen_prefix, core.infra_id
```